### PR TITLE
Support upload of any binary content of the blob during the initial summary upload at routerlicious driver

### PR DESCRIPTION
**Pull request by NicholasCouri**
## Description
The [issue 15969](https://github.com/microsoft/FluidFramework/issues/15969) details the current problem when uploading binary content blobs during the first summary upload. 

In short, the function **convertSummaryToCreateNewSummary** traverses the **ISummaryTree** and encodes the **ISummaryBlob** binary content into string using 'utf-8' encoding which only works in case the original binary content is utf-8 compatible. Not only that but when encoding we re-use the same interface to support the new content (**ISummaryTree** yielding an **ISummaryTree**). 

Ideally, we want the protocol to make use of distinct formats for decoded and encoded summaries.

FRS's historian endpoint already uses a distinct format for encoded summaries, which does correctly preserve encodings. 
The idea behind this PR is to make use of convertSummaryTreeToWholeSummaryTree, which is the encoded format mentioned above, and simply convert back to ISummaryTree on the endpoint with minor changes on the existing processing.  In order to accomplish this, the current PR generates a new method to do the conversion(**convertWholeSummaryTreeToSummaryTree**). 

ps. This is an initial draft PR and I will probably need to split and bump the @fluidframework/server-services-client version before proceeding. That or simply duplicating the convertWholeSummaryTreeToSummaryTree and stripping it later on.

### CODE REVIEW

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 1]
```diff
@@ -162,9 +163,13 @@ export function create(
 				? uuid()
 				: (request.body.id as string) || uuid();
 
+			Lumberjack.info(
+				`Whole summary length= ${JSON.stringify(request.body.summary).length}.`,
+			);
 			// Summary information
-			const summary = request.body.summary;
+			const summary = convertWholeSummaryTreeToSummaryTree(request.body.summary);
```
Comment (vladsud) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:170]: If I got you right, you are changing protocol between client & service. There needs to be some consideration for back compatibility. I.e. old client talking to new server, and new client talking to old server. We can eliminate one pair by timing properly release cadence (i.e. if we ship service changes first and wait for it to propagate through the system, then new client - old service pair is eliminated), but there is no way to eliminate both pairs.
Given that clients take > 3 years to move off old bits, I'd suggest adding auto-detection mechanism in service code to work with both formats.

Q: Does our tests capture that need? If not, feels like we need to remedy this somehow.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 2]
```diff
@@ -162,9 +163,13 @@ export function create(
 				? uuid()
 				: (request.body.id as string) || uuid();
 
+			Lumberjack.info(
+				`Whole summary length= ${JSON.stringify(request.body.summary).length}.`,
```
Comment (vladsud) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:167]: This is too expensive operation to have as logging statement. Same for below.

File: server/routerlicious/packages/services-client/src/storageUtils.ts [hunk 3]
```diff
@@ -214,3 +216,42 @@ export function convertWholeFlatSummaryToSnapshotTreeAndBlobs(
...
+/**
+ * Converts existing IWholeSummaryTree to ISummaryTree
+ * @param wholeSummaryTree - wholeSummaryTree used on the payload for creating and uploading a document.
+ * @returns Summary tree to be used when creating a new document.
+ */
+export function convertWholeSummaryTreeToSummaryTree(
+	wholeSummaryTree: IWholeSummaryTree,
+): ISummaryTree {
+	const tree: { [path: string]: SummaryObject } = {};
+	if (wholeSummaryTree.entries) {
```
Comment (vladsud) [inline server/routerlicious/packages/services-client/src/storageUtils.ts:229]: this IF seems to be redundant.

File: server/routerlicious/packages/services-client/src/storageUtils.ts [hunk 4]
```diff
@@ -214,3 +216,42 @@ export function convertWholeFlatSummaryToSnapshotTreeAndBlobs(
 		sequenceNumber,
 	};
 }
+
+/**
+ * Converts existing IWholeSummaryTree to ISummaryTree
+ * @param wholeSummaryTree - wholeSummaryTree used on the payload for creating and uploading a document.
+ * @returns Summary tree to be used when creating a new document.
+ */
+export function convertWholeSummaryTreeToSummaryTree(
```
Comment (vladsud) [inline server/routerlicious/packages/services-client/src/storageUtils.ts:225]: As a general concept, a summary could have a handle type - a reference to a sub-tree in previously submitted summary.
This is not supported by this function, so it's not generic function (to convert any summary).
We probably want to name it appropriately such that someone accidently does not use it for that.
Or better - I'd create a new interface for that, that excludes IWholeSummaryTreeHandleEntry type.

File: server/tinylicious/src/services/inMemorycollection.ts [hunk 5]
```diff
@@ -83,12 +83,26 @@ export class Collection<T> implements ICollection<T> {
 		});
 	}
 
+	private removeOneInternal(value: any): void {
```
Comment (vladsud) [inline server/tinylicious/src/services/inMemorycollection.ts:86]: Are these changes related to your PR? I'm not sure what's the relationship, and what these changes do.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts (hunk 1 above)
Comment (NicholasCouri) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:170]: Great point! I can easily add a flag to the protocol, indicating the new format and only convert if the flag is sent.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts (hunk 2 above)
Comment (NicholasCouri) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:167]: yes, I was planning on removing it, this is just to make sure the new code is being hit.

File: server/routerlicious/packages/services-client/src/storageUtils.ts (hunk 4 above)
Comment (NicholasCouri) [reply server/routerlicious/packages/services-client/src/storageUtils.ts:225]: I will rename it.

File: server/tinylicious/src/services/inMemorycollection.ts (hunk 5 above)
Comment (NicholasCouri) [reply server/tinylicious/src/services/inMemorycollection.ts:86]: yes, I will create a new PR for it as these methods are throwing on  my local tests.

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts [hunk 6]
```diff
@@ -150,11 +152,17 @@ export class RouterliciousDocumentServiceFactory implements IDocumentServiceFact
...
+							undefined,
+							appSummary,
+							"",
+							"",
+						),
 						sequenceNumber: documentAttributes.sequenceNumber,
 						values: quorumValues,
 						enableDiscovery: this.driverPolicies.enableDiscovery,
 						generateToken: this.tokenProvider.documentPostCreateCallback !== undefined,
+						enableWholeSummaryTreeOnFirstSummary: true,
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:165]: I'd suggest shorter name.

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 7]
```diff
@@ -162,9 +163,16 @@ export function create(
 				? uuid()
 				: (request.body.id as string) || uuid();
 
+			Lumberjack.info(
+				`Whole summary on First Summary ${request.body.enableWholeSummaryTreeOnFirstSummary}.`,
+			);
 			// Summary information
-			const summary = request.body.summary;
+			const summary =
+				request.body.enableWholeSummaryTreeOnFirstSummary === true
```
Comment (vladsud) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:171]: You can do shorter version (i.e. remove === true):
request.body.enableWholeSummaryTreeOnFirstSummary ? caseA : caseB

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts [hunk 8]
```diff
@@ -162,9 +163,16 @@ export function create(
 				? uuid()
 				: (request.body.id as string) || uuid();
 
+			Lumberjack.info(
+				`Whole summary on First Summary ${request.body.enableWholeSummaryTreeOnFirstSummary}.`,
+			);
 			// Summary information
-			const summary = request.body.summary;
+			const summary =
```
Comment (vladsud) [inline server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:170]: This PR is great for demonstration purposes. But I believe old & new way of doing things is incompatible, and thus server changes needs to be checked in separately, and fully deploy to 100% prod before we can make client changes.

File: server/tinylicious/src/routes/ordering/documents.ts [hunk 9]
```diff
@@ -35,8 +39,13 @@ export function create(storage: IDocumentStorage): Router {
 		const tenantId = getParam(request.params, "tenantId");
 		const id = request.body.id || uuid();
 
-		// Summary information
-		const summary = request.body.summary;
+		const summary =
+			request.body.enableWholeSummaryTreeOnFirstSummary === true
```
Comment (vladsud) [inline server/tinylicious/src/routes/ordering/documents.ts:43]: Same comment RE shorter way of expressing same intention as above

File: server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts (hunk 8 above)
Comment (znewton) [reply server/routerlicious/packages/routerlicious-base/src/alfred/routes/api/documents.ts:170]: Or we need it to be configurable (enable/disable) on the driver side so that we can have a follow-up PR to enable this in a specific driver version.

File: packages/drivers/routerlicious-driver/src/documentServiceFactory.ts (hunk 6 above)
Comment (NicholasCouri) [reply packages/drivers/routerlicious-driver/src/documentServiceFactory.ts:165]: The other one I had in mind was enableAnyBinaryBlobOnFirstSummary but it is not as short as I'd like,
